# Data Cleaning

## 1. Import libraries

In [71]:
import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import numpy as np 
import seaborn as sns 
import matplotlib.pyplot as plt 
import os
from sklearn.model_selection import train_test_split


## 2. Import data

In [72]:
PROJECT_DIR = r"/Users/sudhirjoon/Library/Mobile Documents/com~apple~CloudDocs/Uni_Mannheim/Nitish/Project_4_CLASSIFICATION WITH AN ACADEMIC SUCCESS DATASET"
DATA_DIR = "Data"

def get_data(name):
    file_name = f"{name}.csv"
    file_path = os.path.join(PROJECT_DIR, DATA_DIR,file_name)
    return pd.read_csv(file_path)


In [73]:
train = get_data('train')
test = get_data('test')
display(train.head())
display(test.head())

id  Marital status  Application mode  Application order  Course  \
0   0               1                 1                  1    9238   
1   1               1                17                  1    9238   
2   2               1                17                  2    9254   
3   3               1                 1                  3    9500   
4   4               1                 1                  2    9500   

   Daytime/evening attendance  Previous qualification  \
0                           1                       1   
1                           1                       1   
2                           1                       1   
3                           1                       1   
4                           1                       1   

   Previous qualification (grade)  Nacionality  Mother's qualification  \
0                           126.0            1                       1   
1                           125.0            1                      19   
2                           137.0            1                       3   
3                           131.0            1                      19   
4                           132.0            1                      19   

   Father's qualification  Mother's occupation  Father's occupation  \
0                      19                    5                    5   
1                      19                    9                    9   
2                      19                    2                    3   
3                       3                    3                    2   
4                      37                    4                    9   

   Admission grade  Displaced  Educational special needs  Debtor  \
0            122.6          0                          0       0   
1            119.8          1                          0       0   
2            144.7          0                          0       0   
3            126.1          1                          0       0   
4            120.1          1                          0       0   

   Tuition fees up to date  Gender  Scholarship holder  Age at enrollment  \
0                        1       0                   1                 18   
1                        1       0                   0                 18   
2                        1       1                   0                 18   
3                        1       0                   1                 18   
4                        1       0                   0                 18   

   International  Curricular units 1st sem (credited)  \
0              0                                    0   
1              0                                    0   
2              0                                    0   
3              0                                    0   
4              0                                    0   

   Curricular units 1st sem (enrolled)  \
0                                    6   
1                                    6   
2                                    6   
3                                    7   
4                                    7   

   Curricular units 1st sem (evaluations)  \
0                                       6   
1                                       8   
2                                       0   
3                                       9   
4                                      12   

   Curricular units 1st sem (approved)  Curricular units 1st sem (grade)  \
0                                    6                         14.500000   
1                                    4                         11.600000   
2                                    0                          0.000000   
3                                    7                         12.591250   
4                                    6                         12.933333   

   Curricular units 1st sem (without evaluations)  \
0                                               0   
1                                               0   
2           

id  Marital status  Application mode  Application order  Course  \
0  76518               1                 1                  1    9500   
1  76519               1                 1                  1    9238   
2  76520               1                 1                  1    9238   
3  76521               1                44                  1    9147   
4  76522               1                39                  1    9670   

   Daytime/evening attendance  Previous qualification  \
0                           1                       1   
1                           1                       1   
2                           1                       1   
3                           1                      39   
4                           1                       1   

   Previous qualification (grade)  Nacionality  Mother's qualification  \
0                           141.0            1                       3   
1                           128.0            1                       1   
2                           118.0            1                       1   
3                           130.0            1                       1   
4                           110.0            1                       1   

   Father's qualification  Mother's occupation  Father's occupation  \
0                       1                    2                    3   
1                      19                    4                    5   
2                      19                    4                    9   
3                      19                    3                    3   
4                      37                    5                    5   

   Admission grade  Displaced  Educational special needs  Debtor  \
0            152.1          0                          0       0   
1            116.5          0                          0       0   
2            114.2          0                          0       0   
3            130.0          0                          0       0   
4            106.0          1                          0       0   

   Tuition fees up to date  Gender  Scholarship holder  Age at enrollment  \
0                        0       0                   0                 18   
1                        1       0                   0                 19   
2                        1       0                   1                 18   
3                        1       0                   1                 23   
4                        1       0                   0                 26   

   International  Curricular units 1st sem (credited)  \
0              0                                    0   
1              0                                    0   
2              0                                    0   
3              0                                    2   
4              0                                    0   

   Curricular units 1st sem (enrolled)  \
0                                    7   
1                                    6   
2                                    6   
3                                    6   
4                                    6   

   Curricular units 1st sem (evaluations)  \
0                                       0   
1                                       7   
2                                      11   
3                                      15   
4                                       9   

   Curricular units 1st sem (approved)  Curricular units 1st sem (grade)  \
0                                    0                          0.000000   
1                                    6                         14.857143   
2                                    6                         12.000000   
3                                    5                         11.500000   
4                                    3                         11.000000   

   Curricular units 1st sem (without evaluations)  \
0                                               0   
1                                               0 

In [74]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76518 entries, 0 to 76517
Data columns (total 38 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   id                                              76518 non-null  int64  
 1   Marital status                                  76518 non-null  int64  
 2   Application mode                                76518 non-null  int64  
 3   Application order                               76518 non-null  int64  
 4   Course                                          76518 non-null  int64  
 5   Daytime/evening attendance                      76518 non-null  int64  
 6   Previous qualification                          76518 non-null  int64  
 7   Previous qualification (grade)                  76518 non-null  float64
 8   Nacionality                                     76518 non-null  int64  
 9   Mother's qualification                 

In [75]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51012 entries, 0 to 51011
Data columns (total 37 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   id                                              51012 non-null  int64  
 1   Marital status                                  51012 non-null  int64  
 2   Application mode                                51012 non-null  int64  
 3   Application order                               51012 non-null  int64  
 4   Course                                          51012 non-null  int64  
 5   Daytime/evening attendance                      51012 non-null  int64  
 6   Previous qualification                          51012 non-null  int64  
 7   Previous qualification (grade)                  51012 non-null  float64
 8   Nacionality                                     51012 non-null  int64  
 9   Mother's qualification                 

## 3. Preliminary Analysis

In [76]:
#checking for NAs
print(train.isna().sum().sum())
print(test.isna().sum().sum())

0
0


In [77]:
#checking for dupplicates
print(train.duplicated().sum())
print(test.duplicated().sum())

0
0


***Obeservations***
- There are `no duplicated` records 
- There are `no NaN values` present in the train dataset
- There are `76518 rows` and `31 columns` present in the dataset
- 30 columns are of numeric type and target column is object type

In [78]:
train.columns

Index(['id', 'Marital status', 'Application mode', 'Application order',
       'Course', 'Daytime/evening attendance', 'Previous qualification',
       'Previous qualification (grade)', 'Nacionality',
       'Mother's qualification', 'Father's qualification',
       'Mother's occupation', 'Father's occupation', 'Admission grade',
       'Displaced', 'Educational special needs', 'Debtor',
       'Tuition fees up to date', 'Gender', 'Scholarship holder',
       'Age at enrollment', 'International',
       'Curricular units 1st sem (credited)',
       'Curricular units 1st sem (enrolled)',
       'Curricular units 1st sem (evaluations)',
       'Curricular units 1st sem (approved)',
       'Curricular units 1st sem (grade)',
       'Curricular units 1st sem (without evaluations)',
       'Curricular units 2nd sem (credited)',
       'Curricular units 2nd sem (enrolled)',
       'Curricular units 2nd sem (evaluations)',
       'Curricular units 2nd sem (approved)',
       'Curricular units

In [79]:
def rename_columns(data):
    l1 = data.columns.to_list()
    new_columns = ['_'.join(i.split()).lower() for i in l1]
    data.columns = new_columns
    return data.columns


In [80]:
rename_columns(train)
rename_columns(test)

Index(['id', 'marital_status', 'application_mode', 'application_order',
       'course', 'daytime/evening_attendance', 'previous_qualification',
       'previous_qualification_(grade)', 'nacionality',
       'mother's_qualification', 'father's_qualification',
       'mother's_occupation', 'father's_occupation', 'admission_grade',
       'displaced', 'educational_special_needs', 'debtor',
       'tuition_fees_up_to_date', 'gender', 'scholarship_holder',
       'age_at_enrollment', 'international',
       'curricular_units_1st_sem_(credited)',
       'curricular_units_1st_sem_(enrolled)',
       'curricular_units_1st_sem_(evaluations)',
       'curricular_units_1st_sem_(approved)',
       'curricular_units_1st_sem_(grade)',
       'curricular_units_1st_sem_(without_evaluations)',
       'curricular_units_2nd_sem_(credited)',
       'curricular_units_2nd_sem_(enrolled)',
       'curricular_units_2nd_sem_(evaluations)',
       'curricular_units_2nd_sem_(approved)',
       'curricular_units

In [81]:
train.head()

id  marital_status  application_mode  application_order  course  \
0   0               1                 1                  1    9238   
1   1               1                17                  1    9238   
2   2               1                17                  2    9254   
3   3               1                 1                  3    9500   
4   4               1                 1                  2    9500   

   daytime/evening_attendance  previous_qualification  \
0                           1                       1   
1                           1                       1   
2                           1                       1   
3                           1                       1   
4                           1                       1   

   previous_qualification_(grade)  nacionality  mother's_qualification  \
0                           126.0            1                       1   
1                           125.0            1                      19   
2                           137.0            1                       3   
3                           131.0            1                      19   
4                           132.0            1                      19   

   father's_qualification  mother's_occupation  father's_occupation  \
0                      19                    5                    5   
1                      19                    9                    9   
2                      19                    2                    3   
3                       3                    3                    2   
4                      37                    4                    9   

   admission_grade  displaced  educational_special_needs  debtor  \
0            122.6          0                          0       0   
1            119.8          1                          0       0   
2            144.7          0                          0       0   
3            126.1          1                          0       0   
4            120.1          1                          0       0   

   tuition_fees_up_to_date  gender  scholarship_holder  age_at_enrollment  \
0                        1       0                   1                 18   
1                        1       0                   0                 18   
2                        1       1                   0                 18   
3                        1       0                   1                 18   
4                        1       0                   0                 18   

   international  curricular_units_1st_sem_(credited)  \
0              0                                    0   
1              0                                    0   
2              0                                    0   
3              0                                    0   
4              0                                    0   

   curricular_units_1st_sem_(enrolled)  \
0                                    6   
1                                    6   
2                                    6   
3                                    7   
4                                    7   

   curricular_units_1st_sem_(evaluations)  \
0                                       6   
1                                       8   
2                                       0   
3                                       9   
4                                      12   

   curricular_units_1st_sem_(approved)  curricular_units_1st_sem_(grade)  \
0                                    6                         14.500000   
1                                    4                         11.600000   
2                                    0                          0.000000   
3                                    7                         12.591250   
4                                    6                         12.933333   

   curricular_units_1st_sem_(without_evaluations)  \
0                                               0   
1                                               0   
2           

In [82]:
# unique elements in each column
train.nunique()

id                                                76518
marital_status                                        6
application_mode                                     22
application_order                                     8
course                                               19
daytime/evening_attendance                            2
previous_qualification                               21
previous_qualification_(grade)                      110
nacionality                                          18
mother's_qualification                               35
father's_qualification                               39
mother's_occupation                                  40
father's_occupation                                  56
admission_grade                                     668
displaced                                             2
educational_special_needs                             2
debtor                                                2
tuition_fees_up_to_date                         

## 4. Creating train and Validation set 

In [83]:
X = train.drop(columns = ['id','target'], axis= 1)
y = train.target.copy()

In [84]:
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)
X_test = test.drop(columns = 'id')


print(X_train.shape, y_train.shape)
print(X_val.shape ,  y_val.shape)
print(X_test.shape)

(61214, 36) (61214,)
(15304, 36) (15304,)
(51012, 36)


## 4. Export train, test and validation set

In [85]:
PROJECT_DIR = r"/Users/sudhirjoon/Library/Mobile Documents/com~apple~CloudDocs/Uni_Mannheim/Nitish/Project_4_CLASSIFICATION WITH AN ACADEMIC SUCCESS DATASET"
DATA_DIR = "Data"

def export_data(X, y = None, name = "data"):
    file_name = f"{name}.csv"
    file_path = os.path.join(PROJECT_DIR, DATA_DIR,file_name)
    
    # Combine X and y if y is provided
    if y is not None:
        combined_data = X.join(y)
    else:
        combined_data = X

    combined_data.to_csv(file_path,index = False)
    return pd.read_csv(file_path).head()

In [86]:
export_data(X_train, y_train, "train_clean")

marital_status  application_mode  application_order  course  \
0               1                44                  1    9085   
1               1                17                  2    9254   
2               1                 1                  1    9773   
3               1                 1                  1     171   
4               1                39                  1    9238   

   daytime/evening_attendance  previous_qualification  \
0                           1                      39   
1                           1                       1   
2                           1                       1   
3                           1                       1   
4                           1                      19   

   previous_qualification_(grade)  nacionality  mother's_qualification  \
0                           150.0            1                      37   
1                           127.0            1                       1   
2                           140.0            1                      37   
3                           110.0            1                      19   
4                           133.1            1                      34   

   father's_qualification  mother's_occupation  father's_occupation  \
0                      37                    9                    8   
1                       1                    4                    8   
2                      38                    9                    9   
3                       2                    5                    2   
4                      38                    0                    4   

   admission_grade  displaced  educational_special_needs  debtor  \
0            150.0          1                          0       0   
1            122.2          1                          0       0   
2            132.1          1                          0       0   
3            117.7          0                          0       0   
4            100.0          0                          0       0   

   tuition_fees_up_to_date  gender  scholarship_holder  age_at_enrollment  \
0                        1       0                   1                 21   
1                        1       1                   0                 20   
2                        1       0                   1                 19   
3                        1       1                   0                 18   
4                        1       1                   0                 27   

   international  curricular_units_1st_sem_(credited)  \
0              0                                    1   
1              0                                    0   
2              0                                    0   
3              0                                    0   
4              0                                    0   

   curricular_units_1st_sem_(enrolled)  \
0                                    6   
1                                    6   
2                                    6   
3                                    0   
4                                    6   

   curricular_units_1st_sem_(evaluations)  \
0                                       8   
1                                      10   
2                                       9   
3                                       0   
4                                      12   

   curricular_units_1st_sem_(approved)  curricular_units_1st_sem_(grade)  \
0                                    6                         12.666667   
1                                    3                         11.666667   
2                                    4                         11.500000   
3                                    0                          0.000000   
4                                    0                          0.000000   

   curricular_units_1st_sem_(without_evaluations)  \
0                                               0   
1                                               0   
2                                   

In [87]:
export_data(X_val, y_val, "val")

marital_status  application_mode  application_order  course  \
0               1                 1                  2    9500   
1               4                39                  1    9853   
2               1                17                  3    9773   
3               1                 1                  2     171   
4               1                 1                  1    9238   

   daytime/evening_attendance  previous_qualification  \
0                           1                       1   
1                           1                      12   
2                           1                       1   
3                           1                       1   
4                           1                       1   

   previous_qualification_(grade)  nacionality  mother's_qualification  \
0                           133.0            1                       1   
1                           133.1            1                      34   
2                           133.0            1                      38   
3                           133.1            1                      37   
4                           124.0            1                       1   

   father's_qualification  mother's_occupation  father's_occupation  \
0                      37                    9                    9   
1                      34                    0                    0   
2                      38                    5                    5   
3                      37                    5                    5   
4                      19                    9                    4   

   admission_grade  displaced  educational_special_needs  debtor  \
0            126.3          1                          0       0   
1            108.4          0                          0       0   
2            126.1          1                          0       0   
3            130.0          0                          0       0   
4            110.3          0                          0       0   

   tuition_fees_up_to_date  gender  scholarship_holder  age_at_enrollment  \
0                        1       0                   1                 20   
1                        1       0                   0                 51   
2                        1       1                   0                 18   
3                        1       0                   0                 21   
4                        1       0                   0                 19   

   international  curricular_units_1st_sem_(credited)  \
0              0                                    0   
1              0                                    0   
2              0                                    0   
3              0                                    0   
4              0                                    0   

   curricular_units_1st_sem_(enrolled)  \
0                                    8   
1                                    6   
2                                    6   
3                                    0   
4                                    6   

   curricular_units_1st_sem_(evaluations)  \
0                                       8   
1                                       7   
2                                       8   
3                                       0   
4                                      13   

   curricular_units_1st_sem_(approved)  curricular_units_1st_sem_(grade)  \
0                                    7                              13.0   
1                                    2                              11.0   
2                                    6                              12.0   
3                                    0                               0.0   
4                                    5                              11.8   

   curricular_units_1st_sem_(without_evaluations)  \
0                                               0   
1                                               0   
2                                   

In [88]:
export_data(X_test, None , "test_clean")

marital_status  application_mode  application_order  course  \
0               1                 1                  1    9500   
1               1                 1                  1    9238   
2               1                 1                  1    9238   
3               1                44                  1    9147   
4               1                39                  1    9670   

   daytime/evening_attendance  previous_qualification  \
0                           1                       1   
1                           1                       1   
2                           1                       1   
3                           1                      39   
4                           1                       1   

   previous_qualification_(grade)  nacionality  mother's_qualification  \
0                           141.0            1                       3   
1                           128.0            1                       1   
2                           118.0            1                       1   
3                           130.0            1                       1   
4                           110.0            1                       1   

   father's_qualification  mother's_occupation  father's_occupation  \
0                       1                    2                    3   
1                      19                    4                    5   
2                      19                    4                    9   
3                      19                    3                    3   
4                      37                    5                    5   

   admission_grade  displaced  educational_special_needs  debtor  \
0            152.1          0                          0       0   
1            116.5          0                          0       0   
2            114.2          0                          0       0   
3            130.0          0                          0       0   
4            106.0          1                          0       0   

   tuition_fees_up_to_date  gender  scholarship_holder  age_at_enrollment  \
0                        0       0                   0                 18   
1                        1       0                   0                 19   
2                        1       0                   1                 18   
3                        1       0                   1                 23   
4                        1       0                   0                 26   

   international  curricular_units_1st_sem_(credited)  \
0              0                                    0   
1              0                                    0   
2              0                                    0   
3              0                                    2   
4              0                                    0   

   curricular_units_1st_sem_(enrolled)  \
0                                    7   
1                                    6   
2                                    6   
3                                    6   
4                                    6   

   curricular_units_1st_sem_(evaluations)  \
0                                       0   
1                                       7   
2                                      11   
3                                      15   
4                                       9   

   curricular_units_1st_sem_(approved)  curricular_units_1st_sem_(grade)  \
0                                    0                          0.000000   
1                                    6                         14.857143   
2                                    6                         12.000000   
3                                    5                         11.500000   
4                                    3                         11.000000   

   curricular_units_1st_sem_(without_evaluations)  \
0                                               0   
1                                               0   
2                                   